### PR TITLE
refactor: adding/using a semver class instead of version strings

### PR DESCRIPTION
**Pull request by porcellus**
## Summary of change

Adding/using a SemVer class instead of version strings. This will used in #566 to enable version ranges instead of equality checks for CDI versions

## Related issues

- Link to issue1 here
- Link to issue1 here

## Test Plan

Existing tests should cover everything

## Documentation changes

N/A

## Checklist for important updates

- [x] Changelog has been updated
    - [x] If there are any db schema changes, mention those changes clearly
- [x] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [x] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
    - In `build.gradle`
- [x] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [x] Had installed and ran the pre-commit hook
- [x] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [x] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.

### CODE REVIEW

File: src/main/java/io/supertokens/utils/SemVer.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            return false;
+
+        if(this.getClass() != that.getClass())
+            return false;
+        return this.compareTo((SemVer) that) == 0;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(version);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/utils/SemVer.java:81]:
```suggestion
        return version.hashCode();
```

File: src/main/java/io/supertokens/utils/SemVer.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        return 0;
+    }
+
+    @Override public boolean equals(Object that) {
+        if(this == that)
+            return true;
+        if(that == null)
+            return false;
+
+        if(this.getClass() != that.getClass())
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/utils/SemVer.java:74]: Use instanceof instead

File: src/main/java/io/supertokens/webserver/api/emailpassword/SignUpAPI.java [hunk 3]
```diff
@@ -70,9 +70,6 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
             JsonObject result = new JsonObject();
             result.addProperty("status", "OK");
             JsonObject userJson = new JsonParser().parse(new Gson().toJson(user)).getAsJsonObject();
-            if (super.getVersionFromRequest(req).equals("2.4")) {
-                userJson.remove("timeJoined");
-            }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/emailpassword/SignUpAPI.java:75]: why is this removed?

File: src/main/java/io/supertokens/webserver/api/emailpassword/UserAPI.java [hunk 4]
```diff
@@ -103,9 +103,6 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IO
                 JsonObject result = new JsonObject();
                 result.addProperty("status", "OK");
                 JsonObject userJson = new JsonParser().parse(new Gson().toJson(user)).getAsJsonObject();
-                if (super.getVersionFromRequest(req).equals("2.4")) {
-                    userJson.remove("timeJoined");
-                }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/emailpassword/UserAPI.java:108]: why is this removed?

File: src/main/java/io/supertokens/utils/SemVer.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    public SemVer(String version) {
+        if(version == null)
+            throw new IllegalArgumentException("Version can not be null");
+        if(!version.matches("[0-9]+(\\.[0-9]+)*"))
+            throw new IllegalArgumentException("Invalid version format");
+        this.version = version;
+    }
+
+    @Override public int compareTo(SemVer that) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/utils/SemVer.java:49]: - You should add extensive unit tests for this.
- Also, this function can be very easily misused.. it's not clear what it means if this function returns `-1`. Instead, i would prefer to have a function like isGreaterThan, isLesserThan, isEqualTo

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/webserver/api/emailpassword/SignUpAPI.java (hunk 3 above)
Comment (porcellus) [reply src/main/java/io/supertokens/webserver/api/emailpassword/SignUpAPI.java:75]: We no longer support `2.4`, (2.7 is the lowest) right now.

File: src/main/java/io/supertokens/webserver/api/emailpassword/UserAPI.java (hunk 4 above)
Comment (porcellus) [reply src/main/java/io/supertokens/webserver/api/emailpassword/UserAPI.java:108]: We no longer support `2.4`, (2.7 is the lowest) right now.

File: src/main/java/io/supertokens/utils/SemVer.java (hunk 5 above)
Comment (porcellus) [reply src/main/java/io/supertokens/utils/SemVer.java:49]: This is the default `Comparable` interface from java that I think we should definitely implement. I'll add those functions in the other core PR since they will be used there. I debated adding them here and replacing equality checks, but I wanted to make this PR with minimal logic changes.
Btw, we basically only need:
- `equals`
- `atLeast` (or `greaterThanOrEqualTo`)
- `betweenInclusive` (for version ranges like `2.9`-`2.18`)
with maybe a `lesserThan` (but not equal) in the future

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/utils/SemVer.java:49]: Yea. Please implement greaterThanOrEqualTo, betweenInclusive and lesserThan

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]
